File: src/main/scala/BasicOOP/Person.java
// src/main/scala/BasicOOP/Person.java

package BasicOOP;

public class Person {
  private String name;
  private int    age;

  public Person(String name, int age) {
    this.name = name;
    this.age  = age;
  }

  public void   setName(String name) { this.name = name; }
  public String getName()            { return this.name; }

  public void setAge(int age) { this.age = age;  }
  public int  getAge()        { return this.age; }
}

File: src/main/scala/progscala3/basicoop/people/Person.scala
// src/main/scala/progscala3/basicoop/people/Person.scala
package progscala3.basicoop.people

case class Person(
  name: String,
  age: Option[Int] = None,
  address: Option[Address] = None)

object Person:
  def apply(name: String): Person = new Person(name)
  def apply(name: String, age: Int): Person = new Person(name, Some(age))
  def apply(name: String, age: Int, address: Address): Person =
    new Person(name, Some(age), Some(address))
  def apply(name: String, address: Address): Person =
    new Person(name, address = Some(address))
